---
title: "Gene co-expression network"
author: "Francisca Samsing"
date: "07/11/2018"
output: html_document
---
https://bioconductor.org/packages/release/bioc/vignettes/CVE/inst/doc/WGCNA_from_TCGA_RNAseq.html

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning = FALSE, fig.align="center")
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
library(knitr)
```

```{r}
#Loading required packages
library(edgeR)
library(limma)
library(WGCNA)
```

```{r, include=FALSE}
# Import raw Counts, re-name and reorder columns 
countdata <- read.csv('Data/read_counts_clean.txt', header = TRUE, sep = '\t',  row.names = 'Geneid')

colnames(countdata) <- c('ISAV24.R1', 'ISAV24.R2', 'ISAV24.R3', 'ISAV6.R1', 'ISAV6.R2', 'ISAV6.R3', 
                      'Control.R1', 'Control.R2', 'Control.R3',
                      'POMV24.R1', 'POMV24.R2', 'POMV24.R3', 'POMV6.R1', 'POMV6.R2', 'POMV6.R3')

countdata2 <- countdata[,c(7,8,9,13,14,15,4,5,6,10,11,12,1,2,3)]

DataGroups <- c('Control', 'Control', 'Control', 'POMV6', 'POMV6', 'POMV6', 
            'ISAV6', 'ISAV6', 'ISAV6', 'POMV24', 'POMV24', 'POMV24', 
            'ISAV24', 'ISAV24', 'ISAV24')

head(countdata2)
```


```{r}
# Create DGE (Differential gene expression) object using EdgeR
dgList <- DGEList(counts=countdata2,group=factor(DataGroups))
y <- dgList # Rename the object to y to keep a clean version of the object (dgList)
row.names(y$counts) = gsub("LOC(\\d)", "\\1", row.names(y$counts)) ## removing the bloody LOCs 
```

```{r}
library(limma)
RNAseq_voom = voom(y)$E
```

```{r}
#transpose matrix to correlate genes in the following
WGCNA_matrix = t(RNAseq_voom[order(apply(RNAseq_voom,1,mad), decreasing = T)[1:2000],])
```

## Construction of co-expression network

```{r}
#similarity measure between gene profiles: biweight midcorrelation
s <- abs(bicor(WGCNA_matrix))
s[1:4,1:4]
```


```{r}
powers = c(c(1:10), seq(from = 12, to=20, by=2))
sft = pickSoftThreshold(WGCNA_matrix, powerVector = powers, verbose = 5)
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab='Soft Threshold (power)',ylab='Scale Free Topology Model Fit,signed R^2',
     type='n', main = paste('Scale independence'));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=1,col='red'); abline(h=0.90,col='red')
```

```{r}
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers,col="red")
```

```{r}
#calculation of adjacency matrix
beta <- 3
a <- s^beta
#dissimilarity measure
w <- 1-a
w[1:4,1:4]

# Turn adjacency into topological overlap
#TOM <- TOMsimilarity(adjacency);
#dissTOM <- 1-TOM
#dissTOM[1:4,1:4]
```

```{r}
# Call the hierarchical clustering function
geneTree <- hclust(as.dist(w), method = "average");

# Plot the resulting clustering tree (dendrogram)
sizeGrWindow(12,9)
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
labels = FALSE, hang = 0.04);
```

```{r}
# We like large modules, so we set the minimum module size relatively high:
minModuleSize <- 30
# Module identification using dynamic tree cut:
dynamicMods <- cutreeDynamic(dendro = geneTree, distM = w,
deepSplit = 2, pamRespectsDendro = FALSE,
minClusterSize = minModuleSize)

table(dynamicMods)
```

```{r}
# Convert numeric lables into colors
dynamicColors <- labels2colors(dynamicMods)
table(dynamicColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)

plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main = "Gene dendrogram and module colors")

```

```{r}
# Calculate eigengenes
MEList = moduleEigengenes(WGCNA_matrix, colors = dynamicColors)
MEs = MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(METree, main = "Clustering of module eigengenes",
xlab = "", sub = "")
MEDissThres <- 0.25
# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")
# Call an automatic merging function
merge <- mergeCloseModules(WGCNA_matrix, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors <- merge$colors
# Eigengenes of the new merg
mergedMEs <- merge$newMEs
```

```{r}
sizeGrWindow(12, 9)
#pdf(file = "Plots/geneDendro-3.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
c("Dynamic Tree Cut", "Merged dynamic"),
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05)
#dev.off()
```

```{r}
# Rename to moduleColors
moduleColors <- mergedColors
# Construct numerical labels corresponding to the colors
colorOrder <- c("grey", standardColors(50))
moduleLabels <- match(moduleColors, colorOrder)-1
MEs <- mergedMEs
```

```{r}
# Transform w distance matrix with a power to make moderately strong connections more visible in the heatmap
plotTOM = w^7
# Set diagonal to NA for a nicer plot
diag(plotTOM) = NA
# Call the plot function
sizeGrWindow(9,9)
TOMplot(plotTOM, geneTree, moduleColors, main = "Network heatmap plot, all genes")
```

```{r}
# Recalculate module eigengenes
MEs <- moduleEigengenes(WGCNA_matrix, moduleColors)$eigengenes
# Isolate weight from the clinical traits
#weight = as.data.frame(datTraits$weight_g);
#names(weight) = "weight"
# Add the weight to existing module eigengenes
#MET = orderMEs(cbind(MEs, weight))
# Plot the relationships among the eigengenes and the trait
sizeGrWindow(5,7.5);
par(cex = 0.9)
plotEigengeneNetworks(MEs, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle
= 90)

```

```{r}
# Plot the dendrogram
sizeGrWindow(6,6)
par(cex = 1.0)
plotEigengeneNetworks(MEs, "Eigengene dendrogram", marDendro = c(0,4,2,0),
plotHeatmaps = FALSE)
# Plot the heatmap matrix (note: this plot will overwrite the dendrogram plot)
par(cex = 1.0)
plotEigengeneNetworks(MEs, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
plotDendrograms = FALSE, xLabelsAngle = 90)
```

## Export networks to cytoscape


```



```{r}
# Select modules
modules <- c("cyan", "greenyellow")
# Select module probes
probes <- colnames(WGCNA_matrix)
inModule <- moduleColors==modules
modProbes <- probes[inModule]
modGenes = ann$GID[match(modProbes, ann$GID)]
# Select the corresponding Topological Overlap
modTOM <- w[inModule, inModule]
```



```{r}
dimnames(modTOM) <- list(modProbes, modProbes)

# Export the network into edge and node list files Cytoscape can read
cyt <- exportNetworkToCytoscape(modTOM,
edgeFile = paste("CytoscapeInput-edges-", paste(modules, collapse="-"), ".txt", sep=""),
nodeFile = paste("CytoscapeInput-nodes-", paste(modules, collapse="-"), ".txt", sep=""),
weighted = TRUE,
threshold = 0.02,
nodeNames = modProbes,
altNodeNames = modGenes,
nodeAttr = moduleColors[inModule])

```















