---
title: "GO Enrichment Analysis: Biological Processes"
author: "Francisca Samsing"
date: "18/12/2018"
output: html_document
---

https://guangchuangyu.github.io/2016/01/go-analysis-using-clusterprofiler/

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning = FALSE, fig.align="center")
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
library(knitr)
```

```{r}
# Load library and search for the organism
library(clusterProfiler)
library(pathview)
library(tidyverse)
library(enrichplot)
```

Build the Salmon Data Base and import the gene universe (background genes from Salmon genome)

```{r}
library(Category)
library(AnnotationHub)
hub <- AnnotationHub()
query(hub, c("salmo salar","orgdb"))
salmodb <- hub[["AH66207"]]
DatPkgFactory(salmodb)
columns(salmodb)

gene_universe <- read.csv('Results/gene_universe.csv')

```

## PATHWAY ANALYSIS FOR POMV 6 HPI

The following code loads ALL differentially expressed genes with an adjusted p-value < 0.05, and then selects those with an absolute logfold change of 2. 

```{r}
POMV6 <- read.csv('Results/ControlvsPOMV6_ALL.csv') # CHECK FOR NAs
#POMV6_SigGeneList <- POMV6[abs(POMV6$logFC) >= 2,] ## up and down regulated 
POMV6_SigGeneList <- POMV6[POMV6$logFC >= 2,] ## upregulated 
#POMV6_SigGeneList <- POMV6[POMV6$logFC <= -2,] ## downregulated 

rownames(POMV6_SigGeneList) <- NULL
POMV6_SigGeneList_logFC <- POMV6_SigGeneList$logFC
names(POMV6_SigGeneList_logFC) <- POMV6_SigGeneList$ENTREZID
POMV6_SigGeneList_logFC <- sort(POMV6_SigGeneList_logFC, decreasing = TRUE)

ego_POMV6 <- enrichGO(gene = names(POMV6_SigGeneList_logFC),
                universe = as.character(gene_universe$ENTREZID),
                keyType = 'ENTREZID',
                OrgDb         = salmodb,
                ont           = "ALL",
                pAdjustMethod = "fdr",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05)

ego_POMV6_summary <- data.frame(ego_POMV6)
#ego_POMV6_REVIGO <- ego_POMV6_summary[,c('ID', 'p.adjust')]
#write.csv(ego_POMV6_summary, file = "Results/ego_POMV6.csv", row.names=FALSE)

head(summary(ego_POMV6))
```


```{r}
barplot(ego_POMV6, showCategory=20, font.size = 10)
dotplot(ego_POMV6, showCategory=20, font.size = 10)
emapplot(ego_POMV6)
cnetplot(ego_POMV6, categorySize="genNUM", foldChange=POMV6_SigGeneList_logFC)

heatplot(ego_POMV6, foldChange=POMV6_SigGeneList_logFC)
```

## PATHWAY ANALYSIS FOR POMV 24 HPI

The following code loads ALL differentially expressed genes with an adjusted p-value < 0.05, and then selects those with an absolute logfold change of 2. 

```{r}
POMV24 <- read.csv('Results/ControlvsPOMV24_ALL.csv') # CHECK FOR NAs
#POMV24_SigGeneList <- POMV24[abs(POMV24$logFC) >= 2,]
POMV24_SigGeneList <- POMV24[POMV24$logFC >= 2,]
#POMV24_SigGeneList <- POMV24[POMV24$logFC <= -2,]

rownames(POMV24_SigGeneList) <- NULL
POMV24_SigGeneList_logFC <- POMV24_SigGeneList$logFC
names(POMV24_SigGeneList_logFC) <- POMV24_SigGeneList$ENTREZID
POMV24_SigGeneList_logFC <- sort(POMV24_SigGeneList_logFC, decreasing = TRUE)

ego_POMV24 <- enrichGO(gene = names(POMV24_SigGeneList_logFC),
                universe = as.character(gene_universe$ENTREZID),
                keyType = 'ENTREZID',
                OrgDb         = salmodb,
                ont           = "ALL",
                pAdjustMethod = "fdr",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05)

ego_POMV24_summary <- data.frame(ego_POMV24)
#ego_POMV24_REVIGO <- ego_POMV24_summary[,c('ID', 'p.adjust')]
#write.csv(ego_POMV24_summary, file = "Results/ego_POMV24.csv", row.names=FALSE)

head(summary(ego_POMV24))
```

```{r}
barplot(ego_POMV24, showCategory=20, font.size = 10)
dotplot(ego_POMV24, showCategory=20, font.size = 10)
emapplot(ego_POMV24, showCategory = 20)
cnetplot(ego_POMV24, categorySize="genNUM", foldChange=POMV24_SigGeneList_logFC, node_label = FALSE)
```

## PATHWAY ANALYSIS FOR ISAV 6 HPI

The following code loads ALL differentially expressed genes with an adjusted p-value < 0.05, and then selects those with an absolute logfold change of 2. 

```{r}
ISAV6 <- read.csv('Results/ControlvsISAV6_ALL.csv') # CHECK FOR NAs
#ISAV6_SigGeneList <- ISAV6[abs(ISAV6$logFC) >= 2,]
ISAV6_SigGeneList <- ISAV6[ISAV6$logFC >= 2,]
#ISAV6_SigGeneList <- ISAV6[ISAV6$logFC <= -2,]

rownames(ISAV6_SigGeneList) <- NULL
ISAV6_SigGeneList_logFC <- ISAV6_SigGeneList$logFC
names(ISAV6_SigGeneList_logFC) <- ISAV6_SigGeneList$ENTREZID
ISAV6_SigGeneList_logFC <- sort(ISAV6_SigGeneList_logFC, decreasing = TRUE)

ego_ISAV6 <- enrichGO(gene = names(ISAV6_SigGeneList_logFC),
                universe = as.character(gene_universe$ENTREZID),
                keyType = 'ENTREZID',
                OrgDb         = salmodb,
                ont           = "ALL",
                pAdjustMethod = "fdr",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05)

ego_ISAV6_summary <- data.frame(ego_ISAV6)
#ego_ISAV6_REVIGO <- ego_ISAV6_summary[,c('ID', 'p.adjust')]
write.csv(ego_ISAV6_summary, file = "Results/ego_ISAV6.csv", row.names=FALSE)

head(summary(ego_ISAV6))
```


```{r}
barplot(ego_ISAV6, showCategory=20, font.size = 10)
dotplot(ego_ISAV6, showCategory=20, font.size = 10)
emapplot(ego_ISAV6, font.size = 10)
cnetplot(ego_ISAV6, categorySize="genNUM", foldChange=POMV6_SigGeneList_logFC, font.size = 10)
```

## PATHWAY ANALYSIS FOR ISAV 24 HPI

The following code loads ALL differentially expressed genes with an adjusted p-value < 0.05, and then selects those with an absolute logfold change of 2. 

```{r}
ISAV24 <- read.csv('Results/ControlvsISAV24_ALL.csv') # CHECK FOR NAs
#ISAV24_SigGeneList <- ISAV24[abs(ISAV24$logFC) >= 2,]
ISAV24_SigGeneList <- ISAV24[ISAV24$logFC >= 2,]
#ISAV24_SigGeneList <- ISAV24[ISAV24$logFC <= -2,]

rownames(ISAV24_SigGeneList) <- NULL
ISAV24_SigGeneList_logFC <- ISAV24_SigGeneList$logFC
names(ISAV24_SigGeneList_logFC) <- ISAV24_SigGeneList$ENTREZID
ISAV24_SigGeneList_logFC <- sort(ISAV24_SigGeneList_logFC, decreasing = TRUE)

ego_ISAV24 <- enrichGO(gene = names(ISAV24_SigGeneList_logFC),
                universe = as.character(gene_universe$ENTREZID),
                keyType = 'ENTREZID',
                OrgDb         = salmodb,
                ont           = "ALL",
                pAdjustMethod = "fdr",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05)

ego_ISAV24_summary <- data.frame(ego_ISAV24)
#ego_ISAV24 <- ego_ISAV24_summary[,c('ID', 'p.adjust')]
write.csv(ego_ISAV24_summary, file = "Results/ego_ISAV24.csv", row.names=FALSE)

head(summary(ego_ISAV24))
```


```{r}
barplot(ego_ISAV24, showCategory=20, font.size = 10)
dotplot(ego_ISAV24, showCategory=20, font.size = 10)
emapplot(ego_ISAV24, font.size = 10)
cnetplot(ego_ISAV24, categorySize="genNUM", foldChange=POMV6_SigGeneList_logFC, font.size = 10)
```

## GO Gene Set Enrichment Analysis

```{r}
gene_list <- list("POMV 6 hpi" = as.character(names(POMV6_SigGeneList_logFC)), "POMV 24 hpi" = as.character(names(POMV24_SigGeneList_logFC)), 
                  "ISAV 6 hpi" = as.character(names(ISAV6_SigGeneList_logFC)), "ISAV 24 hpi" = as.character(names(ISAV24_SigGeneList_logFC)))

ck <- compareCluster(geneCluster = gene_list, 
                     fun = "enrichGO", 
                     ont = "BP", 
                     OrgDb = salmodb, 
                     pvalueCutoff=0.05)

ck_summary <- data.frame(ck)
```

```{r}
## Save to file ## 
dotplot(ck, showCategory=30) +
        labs(#y = "Biological Processes", 
        size = 'Gene ratio', 
        color = 'Adjusted p-value')
```

```{r}

#write.csv(ck_summary, file = "Results/GO_enrichment_UP.csv", row.names = FALSE)

```


## MAKE dotplot using GGPLOT 

```{r}
ck_summary <- arrange(ck_summary, p.adjust)
ck_summary_top <- ck_summary[1:90,]
str(ck_summary_top)

ggplot(ck_summary_top, aes(x = Cluster, y = Description, size = Count)) + 
  geom_point(aes(color = p.adjust), alpha = 0.8) + 
  labs(y = "Biological Processes", 
       size = 'Number of genes', 
       color = 'Adjusted p-value') +
  theme_bw() +
  theme(axis.title.x=element_blank(), 
        axis.text=element_text(colour="black"), 
        panel.border = element_blank()) +
  guides(fill = guide_colourbar(nbin = 2)) +
  scale_color_gradient(low="tomato", high="blue") + 
  scale_size_continuous(range = c(1, 8))

```



































